### PR TITLE
mclapply by default

### PR DESCRIPTION
**Pull request by kirillseva**

### CODE REVIEW

**Comment by robertzk** [comment]
Tests?

File: R/batch.R [hunk 1]
```diff
@@ -25,7 +25,9 @@ batch <- function(
     batchman.verbose = isTRUE(interactive()),
     stop = FALSE,
     retry = 0,
-    sleep = 0
+    sleep = 0,
+    ncores = parallel::detectCores(),
+    parallel = TRUE
```
Comment (peterhurford) [inline R/batch.R]: Don't make it parallel by default.  I'd rather it be an option people can intentionally enable.  People shouldn't be doing parallel things unless they know what they're doing.

File: R/batch.R [hunk 2]
```diff
@@ -38,31 +40,36 @@ batch <- function(
...
+      new_call <- lapply(batch_info, `[[`, 'new_call')
       run_env <- list2env(list(batch_fn = batch_fn))
-      parent.env(run_env) <- parent.frame(find_in_stack(batch_info$keys[[1]]))
-      p <- if (`verbose_set?`()) progress_bar(batch_info$num_batches)
-
-      while (!is.done(new_call)) {
+      parent.env(run_env) <- parent.frame(find_in_stack(batch_info[[1]]$keys[[1]]))
+      p <- if (`verbose_set?`()) progress_bar(ceiling(batch_info[[1]]$num_batches/ncores))
+      apply_method <- if (isTRUE(parallel)) { parallel::mclapply } else { lapply }
+      is.emptyrun <- function(x) identical(x, structure(NULL, emptyrun = TRUE))
```
Comment (peterhurford) [inline R/batch.R]: Can you move this to utils?

File: R/batch.R [hunk 3]
```diff
@@ -38,31 +40,36 @@ batch <- function(
...
+          if (is.done(newcall)) return(structure(NULL, emptyrun = TRUE))
+          if (isTRUE(trycatch)) {
+            iterated_try_catch(
+              eval(newcall, envir = run_env),
+              newcall,
+              run_env,
+              retry
+            )
+          } else { eval(newcall, envir = run_env) }
+        }, mc.cores = ncores, mc.allow.recursive = FALSE)
```
Comment (peterhurford) [inline R/batch.R]: You should move this line to within the `apply_method`.

File: R/batch.R [hunk 4]
```diff
@@ -38,31 +40,36 @@ batch <- function(
...
-            run_env,
-            retry
-          )
-        } else { eval(new_call, envir = run_env) }
-
         if (sleep > 0) { Sys.sleep(sleep) }
-
-        batches <- if (is.no_batches(batches)) batch
-          else combination_strategy(batches, batch)
+        temp_batches  <- temp_batches[vapply(temp_batches, Negate(is.emptyrun), logical(1))]
```
Comment (peterhurford) [inline R/batch.R]: What does this do?

File: tests/testthat/test-batch.R [hunk 5]
```diff
@@ -232,6 +235,22 @@ test_that("it must be more efficient to batch than to execute an O(x^2) function
...
+    combination_strategy = c, size = 5, batchman.verbose = FALSE, parallel = TRUE)
+  batched_sleep_square_l <- batch(sleep_square, "input",
+    combination_strategy = c, size = 5, batchman.verbose = FALSE, parallel = FALSE)
+  require(microbenchmark)
+  speeds <- summary(microbenchmark(times = 1,
+    batched_sleep_square_l(seq(1:20)),
+    batched_sleep_square_p(seq(1:20))
+  ))
+  expect_true(speeds$median[[2]] < speeds$median[[1]])
+})
```
Comment (peterhurford) [inline tests/testthat/test-batch.R]: Nice. :+1:

File: R/batch.R (hunk 1 above)
Comment (kirillseva) [inline R/batch.R]: True that. I'll fix the tests, then flip, then fix it back. apparently need to sacrifice smth for the paralellism

File: R/batch.R (hunk 2 above)
Comment (kirillseva) [inline R/batch.R]: sure

File: tests/testthat/test-batch.R (hunk 5 above)
Comment (kirillseva) [inline tests/testthat/test-batch.R]: Can't wait to use it in real war

**Comment by peterhurford** [comment]
If you want to be a true hero, update `NEWS.md` and document the new feature in the `README`.

**Comment by peterhurford** [comment]
@kirillseva Tests are still failing

**Comment by kirillseva** [comment]
Not any more @peterhurford

File: R/batch.R [hunk 6]
```diff
@@ -25,8 +28,13 @@ batch <- function(
     batchman.verbose = isTRUE(interactive()),
     stop = FALSE,
     retry = 0,
-    sleep = 0
+    sleep = 0,
+    ncores = parallel::detectCores(),
+    parallel = FALSE
```
Comment (tonglu) [inline R/batch.R]: @peterhurford batch by default uses `lapply`

**Pull request merged**